### ISSUE TITLE
[PATCH] sending new style classes via perspective broker

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8956]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| titty reported|
|-|-|
|Trac ID|trac#426|
|Type|enhancement|
|Created|2003-12-10 23:16:00Z|

Attachments:

* [jelly.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/628/62866ab415331d03c03849c24f528450f246e375/b38108de5b30ecc73138a302cdf5d29b26cbe02f.patch) (1978 bytes) - added by matthew on 2005-04-13 10:55:15Z - 
* [test_jelly.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/628/62866ab415331d03c03849c24f528450f246e375/28d10bf5f977f6fb4dc3ed9930f7c2b4a0e04f80.patch) (1613 bytes) - added by matthew on 2005-04-13 11:04:36Z - 
* [test_jelly.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/628/62866ab415331d03c03849c24f528450f246e375/00ec480a57346f99a80b30cd178aff3f2a7a7cbe.patch) (2114 bytes) - added by matthew on 2005-04-18 12:08:52Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__426 426
type__enhancement enhancement
reporter__titty titty
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1071098160000000 1071098160000000
changetime__1469604403909162 1469604403909162
version__ 
owner__ 
cc__warner cc__titty cc__db3l cc__matthew cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
sending new style classes via perspective broker currently isn't
possible without extra effort. I have attached a short test program
and a patch to jelly.py, which solves that problem.

=========== 
from twisted.spread import pb
from twisted.internet import reactor, threads
from twisted.python import threadable

class Data(pb.Copyable, pb.RemoteCopy, object):
    def __init__(self, s):
        self.s = s

pb.setUnjellyableForClass(Data, Data)
    
class Echoer(pb.Root):
    def remote_echo(self, st):
        print "REMOTE_ECHO:", st, st.s
        return st

if __name__ == '__main__':
    reactor.listenTCP(8789, pb.PBServerFactory(Echoer()))

    factory = pb.PBClientFactory()
    reactor.connectTCP("127.0.0.1", 8789, factory)
    factory.getRootObject().addCallback(lambda r: r.callRemote("echo", Data("foo")))

    reactor.callLater(2, reactor.stop)
    reactor.run()

===================================================================
RCS file: /cvs/Twisted/twisted/spread/jelly.py,v
retrieving revision 1.48
diff -u -r1.48 jelly.py
--- jelly.py    4 Jun 2003 00:18:03 -0000       1.48
+++ jelly.py    10 Dec 2003 12:35:48 -0000
@@ -520,6 +520,9 @@
                 inst = _Dummy() # XXX chomp, chomp
                 inst.__class__ = regClass
                 method = inst.unjellyFor
+            elif isinstance(regClass, type): # new style class
+                inst = object.__new__(regClass) # does not call regClass.__init__
+                method = inst.unjellyFor
             else:
                 method = regClass # this is how it ought to be done
             val = method(self, obj)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Assigning to warner
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="db3l's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|db3l commented|
|-|-|

```
#!html
<pre>
We've been using this patch in our local copy of Twisted (1.1.0 and then 1.3.0)
for a while now without any problems, but I was surprised to notice that it
isn't in the Twisted 2.0 package I just downloaded.  I would think that you'd
want new style class support for pb to be supported by default.  Is there any
intention of rolling this into the main source tree?  Thanks.

-- David
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="matthew's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|matthew commented|
|-|-|

```
#!html
<pre>
This patch fixes it in twisted 2.0 for me.

Apply it to jelly.py
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="matthew's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|matthew commented|
|-|-|

```
#!html
<pre>
New Patch. Modified according to exarkuns suggestions ;)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="matthew's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|matthew commented|
|-|-|

```
#!html
<pre>
Patch for test_jelly.py
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="matthew's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|matthew commented|
|-|-|

```
#!html
<pre>
Tougher test for test_jelly.patch
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
I've applied a modified version of Matthew's patch (just cosmetic changes). Does
this also take care of the case described in the original patch (where a
new-style class was registered with unjellyableRegistry)?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
well, I've attached an test program in my original message.
it still does *not* work (svnversion 13537).
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="matthew's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|matthew commented|
|-|-|

```
#!html
<pre>
titty. What's actually happening to make it be not working?
Also I can't see your attachment
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
you're kidding, aren't you?
the program actually isn't attached, but it's 'inline' in my first message.

Here's the output:
ralf@stronzo:~/bbot/extern/Twisted$ python t.py
Peer will receive following PB traceback:
Traceback (most recent call last):
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/banana.py", line 114,
in gotItem
    self.callExpressionReceived(item)
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/banana.py", line 81,
in callExpressionReceived
    self.expressionReceived(obj)
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/pb.py", line 561, in
expressionReceived
    method(*sexp[1:])
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/pb.py", line 872, in
proto_message
    self._recvMessage(self.localObjectForID, requestID, objectID, message,
answerRequired, netArgs, netKw)
--- &lt;exception caught here> ---
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/pb.py", line 886, in
_recvMessage
    netResult = object.remoteMessageReceived(self, message, netArgs, netKw)
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/flavors.py", line 113,
in remoteMessageReceived
    args = broker.unserialize(args)
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/pb.py", line 822, in
unserialize
    return unjelly(sexp, self.security, None, self)
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/jelly.py", line 902,
in unjelly
    return _Unjellier(taster, persistentLoad, invoker).unjellyFull(sexp)
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/jelly.py", line 535,
in unjellyFull
    o = self.unjelly(obj)
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/jelly.py", line 567,
in unjelly
    ret = thunk(obj[1:])
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/jelly.py", line 650,
in _unjelly_tuple
    if isinstance(self.unjellyInto(l, elem, lst[elem]), NotKnown):
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/jelly.py", line 617,
in unjellyInto
    o = self.unjelly(jel)
  File "/usr/home/ralf/bbot/extern/Twisted/twisted/spread/jelly.py", line 554,
in unjelly
    val = method(self, obj)
exceptions.TypeError: __init__() takes exactly 2 arguments (3 given)
Unhandled error in Deferred:
ralf@stronzo:~/bbot/extern/Twisted$ svnversion .
13538
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
I've adapted titty's test program into test_pb.py, and applied the patch
(r13593). I think this should be done. Please take a look at it and tell me if
there are any problems.

thanks,
 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="matthew's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|matthew commented|
|-|-|

```
#!html
<pre>
Yes, the patch is an improvement, I just changed object.__new__ to
regClass.__new__ so that if the writer of regClass has overridden __new__ it
will be called.

eg.
>>> class X(object):
...  a = 1
...  def __init__(self):
...   self.b = 1
...  def __new__(self):
...   inst = object.__new__(self)
...   inst.c = 1
...   return inst
...
>>> x = X()
>>> x.a
1
>>> x.b
1
>>> x.c
1
>>> x = object.__new__(X)
>>> x.a
1
>>> x.b
Traceback (most recent call last):
  File "&lt;stdin>", line 1, in ?
AttributeError: 'X' object has no attribute 'b'
>>> x.c
Traceback (most recent call last):
  File "&lt;stdin>", line 1, in ?
AttributeError: 'X' object has no attribute 'c'
>>> x = X.__new__(X)
>>> x.a
1
>>> c.b
Traceback (most recent call last):
  File "&lt;stdin>", line 1, in ?
NameError: name 'c' is not defined
>>> x.c
1


Patch Below
----8&lt;-----
Index: jelly.py
===================================================================
--- jelly.py    (revision 13595)
+++ jelly.py    (working copy)
@@ -550,8 +550,8 @@
                 inst.__class__ = regClass
                 method = inst.unjellyFor
             elif isinstance(regClass, type):
-                # object.__new__ does not call regClass.__init__
-                inst = object.__new__(regClass)
+                # regClass.__new__ does not call regClass.__init__
+                inst = regClass.__new__(regClass)
                 method = inst.unjellyFor
             else:
                 method = regClass # this is how it ought to be done
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
I just added your patch, and a test case.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="matthew's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|matthew commented|
|-|-|

```
#!html
<pre>
Here's another patch for test_jelly.py that tests that jelly._newInstance calls
the correct __new__ method.

The below module checks the pb stuff. warner, could you please make it into a
test_pb.py func? :)

-------- 8&lt; --------------
from twisted.spread import pb
from twisted.internet import reactor, threads
from twisted.python import threadable

class Data(pb.Copyable, pb.RemoteCopy, object):
    count = 0
    def __init__(self, s):
        self.s = s
    def __new__(cls, *args, **kwargs):
        Data.count += 1
        return object.__new__(cls, *args, **kwargs)

pb.setUnjellyableForClass(Data, Data)
    
class Echoer(pb.Root):
    def remote_echo(self, st):
        print "REMOTE_ECHO:", st, st.s
        return st

def doecho(r):
    assert Data.count == 0
    dat = Data('foo')
    assert Data.count == 1
    d = r.callRemote("echo", dat)
    d.addCallback(testecho)

def testecho(res):
    reactor.callLater(0, reactor.stop)
    assert res.count == 3, ('Should be 3. Instance 1 is created in doecho.'
                            'Instance 2 is created when the server unjellies '
                            'the argument.'
                            'Instance 3 is created when the client unjellies '
                            'the answer from the server (although its a copy '
                            'of the same object that the client originally sent)')

if __name__ == '__main__':
    reactor.listenTCP(8789, pb.PBServerFactory(Echoer()))
    factory = pb.PBClientFactory()
    reactor.connectTCP("127.0.0.1", 8789, factory)
    factory.getRootObject().addCallback(doecho)
    reactor.run()
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

I think this one is resolved. Can we close it ?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `closed`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|GitHub commented|
|-|-|

In changeset ddbed121626cfcbdf8f13e811af048bc5a9ace04
```
#!CommitTicketReference repository="" revision="ddbed121626cfcbdf8f13e811af048bc5a9ace04"
Merge pull request #426 from twisted/8690-einarfd-conch-ssh

Author: einarfd
Reviewer: rodrigc
Fixes: #8690

Port twisted.conch.factory to Python 3
```
